### ISSUE TITLE
Queue for transfer is missing

### ISSUE THREAD

**Issue by iOvergaard** [umbraco/Umbraco.Deploy.Issues#95]
The option to queue items for transfer ("Queue for transfer") is missing from the tree when rightclicking items. All other Deploy actions like "Restore" are there and working, as well as "Transfer now" in the Save & Publish button.

I can transfer nodes just fine between environments using the available actions, however "Transfer now" is hanging on the progress bar, even though I was able to verify that the deployment went through just fine.

This has been verified on three independent Cloud sites created today.

Reproduction
------------

### Workflows tested

Local -> Live

Development -> Live

### Specifics

Umbraco 9.2.0
Deploy 9.2.0

### Steps to reproduce

1. Create cloud project
2. Clone down locally after creation
3. Create document type
4. Create document & save & publish
5. Right click new document and see that "Queue for transfer" is missing
  a. "Transfer now" on the actual page is tested and working, however the progress bar hangs even though the transfer went through and is visible on the Live site

### Expected result

* "Queue for transfer" should be visible

### Actual result

* "Queue for transfer" is not visible

![image](https://user-images.githubusercontent.com/752371/148408290-242a9031-adad-41e9-86f1-0886728c498a.png)

**Comment by RyuLindow** [comment]
I get the exact same issue with the queue for the transfer button being gone locally, however, the transfer now button works fine together with the progress bar moving and the content appearing on live.

Umbraco 9.2.0
Forms 9.2.0
Deploy 9.2.0

**Comment by AndyButland** [comment]
Thanks for reporting both - I've found the issue.  I was digging around in Deploy 9.2 to try to see if we'd caused this in there, but turns out it's surfaced by an updated in CMS 9.2.  I'm sure this isn't the first time if someone is burnt by it that you'll see when I explain it...

Each menu action item in Umbraco needs to be assigned a single character identifying letter, and if there are two with the same one, it seems that the second one added will be omitted from the menu.  Via [this PR](https://github.com/umbraco/Umbraco-CMS/pull/10450) in a recent update to CMS, a new menu item has been added, with the letter `N`.  Which looks innocuous as [this isn't used by another action in CMS](https://our.umbraco.com/documentation/extending/section-trees/tree-actions)... but it is used by the "Queue For Transfer" action in Deploy.

If I change Deploy to use a different character, and we release a patch for Deploy, that'll then allow both to be used.  Unfortunately though that's not quite all.  These characters are used primarily for assigning permissions, and with a new character for this action, no user will have permission to use it until it's added to their user groups.

Normally when we introduce these actions, and as the linked PR has done, we also have a migration to add the letter to the list of permissions for the appropriate default user groups (e.g. admin and editor).

So we can do that again, but due to this clashing letter use, it's a bit tricky - as it ideally needs to clean up the situation with user groups that have two values for "N" - one original one for Deploy, and one new one for notifications.

So all in all I think we are better fixing this with a CMS patch, that just changes the one picked for notifications to a new, non-clashing letter (for Deploy, please avoid `Q`, `Ψ` and `Ø` too).  If we add a migration we can also find user groups that have been affected by the CMS 9.2. migration - i.e. that have a permission string ending with "N" - and fairly easily replace it with the new letter.

Database migrations are generally limited to minor updates, not patches, but we'd need to break that rule in this instance if we want to ensure projects are auto-upgraded.

**Comment by AndyButland** [comment]
Two other quick points:

- If we end up changing Deploy rather than CMS, we should do so in the version for V8 too, as the same issue will otherwise recur when 8.18 is out.  If we change CMS, we should change in in V8 branch as well as V9 for the same reason.
- When we decide on the new character, let's update [this page](https://our.umbraco.com/documentation/extending/section-trees/tree-actions) on docs with all the ones currently in use for CMS, Deploy and Forms.

**Comment by ronaldbarendse** [comment]
We had an internal discussion about this and came to the following conclusion/action plan:
- Ship a new minor version for Deploy (both v4 and v9) that changes the permission letter for 'Queue for transfer':
  - This new letter shouldn't clash with existing permissions, at least not for CMS and Forms, but preferably also not for 3rd party packages (at least as far as we can check);
  - This new version will contain a database migration that updates the existing `N` permission based on the CMS version:
    - For 8.18+ and 9.2+ we'll update the letter for duplicate occurrences of `N` in the `admin`, `writer` and `editor` roles (as the CMS migration only appends this to those roles) and single occurrences in all other roles. We should add an update note (and possibly log warnings for any changes we make in the migration), in case the new Notifications permission was manually added to other roles (as that would then, mistakenly, be converted to the 'Queue for transfer' permission);
    - For all lower versions we can update the `N` permission in all roles without any issue, so if you update to this new Deploy version before going to 8.18+ or 9.2+, you can be assured all permissions are correct.
- Update the documentation with all permission letters that the CMS, Deploy and Forms use (and possibly for 3rd party packages ass well): https://our.umbraco.com/Documentation/Extending/Section-Trees/tree-actions#user-permission-codes

One thing that I'm still unsure about is how we should handle new installs: if we also change the initial permission assignment in Deploy (in the `AssignDefaultDeployPermissions` migration) to use the new letter, we don't need to run the migration to update the `N` letter again (as that would then only be the one used by the CMS). We can check for the new permission letter and skip the migration when this is already added to one of the default user groups or e.g. by temporarily adding a value to the `umbracoKeyValue` table in the `AssignDefaultDeployPermissions` and check/remove that value in the permission migration again (if all of this makes sense 😇).

**Comment by ronaldbarendse** [comment]
To get an overview of all currently used action letters, I've written a component that logs a message with all required information (in Markdown table format). You can add this to any project, start it up and see which letters are used by the CMS and any of the installed packages:
```c#
using System.Linq;
using Microsoft.Extensions.Logging;
using Umbraco.Cms.Core.Actions;
using Umbraco.Cms.Core.Composing;

public class RegisteredActionsComposer : ComponentComposer<RegisteredActionsComponent>
{ }

public class RegisteredActionsComponent : IComponent
{
    private readonly ILogger<RegisteredActionsComponent> _logger;
    private readonly ActionCollection _actions;

    public RegisteredActionsComponent(ILogger<RegisteredActionsComponent> logger, ActionCollection actions)
    {
        _logger = logger;
        _actions = actions;
    }

    public void Initialize() =>
        _logger.LogInformation(
            "Registered actions:\n|Type|Alias|Letter|CanBePermissionAssigned|\n|-|-|-|-|\n{Actions}",
            string.Join("\n", _actions.Select(x => $"|{x.GetType()}|{x.Alias}|{x.Letter}|{x.CanBePermissionAssigned}|")));

    public void Terminate()
    { }
}
```
### Currently registered actions/permission letters

I've done this on the current `v9/dev` branches of the CMS, Forms and Deploy and compiled the following list:
|Type|Alias|Letter|CanBePermissionAssigned|
|-|-|-|-|
|Umbraco.Cms.Core.Actions.ActionAssignDomain|assignDomain|I|True|
|Umbraco.Cms.Core.Actions.ActionBrowse|browse|F|True|
|Umbraco.Cms.Core.Actions.ActionCopy|copy|O|True|
|Umbraco.Cms.Core.Actions.ActionCreateBlueprintFromContent|createblueprint|‹|True|
|Umbraco.Cms.Core.Actions.ActionDelete|delete|D|True|
|Umbraco.Cms.Core.Actions.ActionMove|move|M|True|
|Umbraco.Cms.Core.Actions.ActionNew|create|C|True|
|Umbraco.Cms.Core.Actions.ActionNotify|notify|N|True|
|Umbraco.Cms.Core.Actions.ActionProtect|protect|P|True|
|Umbraco.Cms.Core.Actions.ActionPublish|publish|U|True|
|Umbraco.Cms.Core.Actions.ActionRestore|restore|V|False|
|Umbraco.Cms.Core.Actions.ActionRights|rights|R|True|
|Umbraco.Cms.Core.Actions.ActionRollback|rollback|K|True|
|Umbraco.Cms.Core.Actions.ActionSort|sort|S|True|
|Umbraco.Cms.Core.Actions.ActionToPublish|sendtopublish|H|True|
|Umbraco.Cms.Core.Actions.ActionUnpublish|unpublish|Z|True|
|Umbraco.Cms.Core.Actions.ActionUpdate|update|A|True|
|Umbraco.Deploy.UI.Actions.ActionDeployRestore|deployRestore|Q|True|
|Umbraco.Deploy.UI.Actions.ActionDeployTreeRestore|deployTreeRestore|?|True|
|Umbraco.Deploy.UI.Actions.ActionPartialRestore|deployPartialRestore|O|True|
|Umbraco.Deploy.UI.Actions.ActionQueueForTransfer|deployQueueForTransfer|N|True|

You'll notice both `Umbraco.Cms.Core.Actions.ActionNotify` and `Umbraco.Deploy.UI.Actions.ActionQueueForTransfer|deployQueueForTransfer` are registered and have the same letter `N`. Because the CMS action is registered first, this will be the one used, as we're using `FirstOrDefault()` here: https://github.com/umbraco/Umbraco-CMS/blob/5bfab13dc5a268714aad2426a2b68ab5561a6407/src/Umbraco.Core/Actions/ActionCollection.cs#L19-L27.

You'll probably also notice that Forms doesn't use the built-in permissions system at all 😇 We're aware of this and will ensure we don't end up with duplicate permission letters again if we revamp Forms permissions.

### Workaround
The 'Queue for transfer' action can be restored by removing the `ActionNotify` type from the registered actions (which will also remove the possibility to [configure content notification in the CMS](https://our.umbraco.com/documentation/Tutorials/Editors-Manual/Tips-and-Tricks/Notifications/)). I'd recommend leaving the duplicate `N` in the `umbracoUserGroup.userGroupDefaultPermissions` database column (and not change this permission), so the fix described in my previous comment can still be correctly applied.

```c#
using Umbraco.Cms.Core.Actions;
using Umbraco.Cms.Core.Composing;
using Umbraco.Cms.Core.DependencyInjection;

public class RemoveActionNotifyComposer : IComposer
{
    public void Compose(IUmbracoBuilder builder)
    {
        builder.Actions().Exclude<ActionNotify>();
    }
}
```

**Comment by ronaldbarendse** [comment]
To following component can be used to get an overview on Umbraco 8:
```c#
using System.Linq;
using Umbraco.Core.Composing;
using Umbraco.Core.Logging;
using Umbraco.Web.Actions;

public class RegisteredActionsComposer : ComponentComposer<RegisteredActionsComponent>
{ }

public class RegisteredActionsComponent : IComponent
{
    private readonly ILogger _logger;
    private readonly ActionCollection _actions;

    public RegisteredActionsComponent(ILogger logger, ActionCollection actions)
    {
        _logger = logger;
        _actions = actions;
    }

    public void Initialize() =>
        _logger.Info<RegisteredActionsComponent>(
            "Registered actions:\n|Type|Alias|Letter|CanBePermissionAssigned|\n|-|-|-|-|\n{Actions}",
            string.Join("\n", _actions.Select(x => $"|{x.GetType()}|{x.Alias}|{x.Letter}|{x.CanBePermissionAssigned}|")));

    public void Terminate()
    { }
}
```

**Comment by AndyButland** [comment]
The fix for this has been tested and merged and will be available in release 9.2.1.
